internal/requests/device_code_response.go

```diff
@@ -4,40 +4,46 @@
 package requests
 
 import (
-	"encoding/json"
 	"time"
 
+	"github.com/AzureAD/microsoft-authentication-library-for-go/internal/json"
 	"github.com/AzureAD/microsoft-authentication-library-for-go/internal/msalbase"
 )
 
 //DeviceCodeResponse represents the HTTP response received from the device code endpoint
 type DeviceCodeResponse struct {
-	BaseResponse    *msalbase.OAuthResponseBase
+	// TODO(jdoak): Ask someone about why BaseResponse doesn't have a tag.
+	// Either it should be encoded and we should tag it or we should tag it
+	// to be omitted on export or private.
+	BaseResponse    msalbase.OAuthResponseBase
 	UserCode        string `json:"user_code"`
 	DeviceCode      string `json:"device_code"`
 	VerificationURL string `json:"verification_url"`
 	ExpiresIn       int    `json:"expires_in"`
 	Interval        int    `json:"interval"`
 	Message         string `json:"message"`
+
+	AdditionalFields map[string]interface{}
 }
 
-// CreateDeviceCodeResponse creates a deviceCodeResponse instance from HTTP response
-func CreateDeviceCodeResponse(responseCode int, responseData string) (*DeviceCodeResponse, error) {
+// CreateDeviceCodeResponse creates a deviceCodeResponse instance from HTTP response.
+// TODO(jdoak): look at responseData and see if we can get it originally from []byte instead.
+func CreateDeviceCodeResponse(responseCode int, responseData string) (DeviceCodeResponse, error) {
+	dcResponse := DeviceCodeResponse{}
 	baseResponse, err := msalbase.CreateOAuthResponseBase(responseCode, responseData)
 	if err != nil {
-		return nil, err
+		return dcResponse, err
 	}
-	dcResponse := &DeviceCodeResponse{}
-	err = json.Unmarshal([]byte(responseData), dcResponse)
-	if err != nil {
-		return nil, err
+
+	if err := json.Unmarshal([]byte(responseData), &dcResponse); err != nil {
+		return dcResponse, err
 	}
 	dcResponse.BaseResponse = baseResponse
 	return dcResponse, nil
 }
 
 //ToDeviceCodeResult converts the DeviceCodeResponse to a DeviceCodeResult
-func (dcr *DeviceCodeResponse) ToDeviceCodeResult(clientID string, scopes []string) *msalbase.DeviceCodeResult {
+func (dcr *DeviceCodeResponse) ToDeviceCodeResult(clientID string, scopes []string) msalbase.DeviceCodeResult {
```

Comment:
Does this still need a pointer receiver?

Reply:
Nope, changed.